packaging/_parser.py

```diff
@@ -54,141 +49,228 @@ def serialize(self) -> str:
 MarkerList = List[Any]
 
 
-class Requirement(NamedTuple):
+class ParsedRequirement(NamedTuple):
     name: str
-    url: str
     extras: List[str]
+    url: str
     specifier: str
-    marker: str
+    marker: Optional[MarkerList]
+
+
+# --------------------------------------------------------------------------------------
+# Recursive descent parser for dependency specifier
+# --------------------------------------------------------------------------------------
+def parse_requirement(source: str) -> ParsedRequirement:
+    return _parse_requirement(Tokenizer(source, rules=DEFAULT_RULES))
+
+
+def _parse_requirement(tokenizer: Tokenizer) -> ParsedRequirement:
+    """
+    requirement = WS? IDENTIFIER WS? extras WS? requirement_details
+    """
+    tokenizer.consume("WS")
+
+    name_token = tokenizer.expect(
+        "IDENTIFIER", expected="package name at the start of dependency specifier"
+    )
+    name = name_token.text
+    tokenizer.consume("WS")
+
+    extras = _parse_extras(tokenizer)
+    tokenizer.consume("WS")
 
+    url, specifier, marker = _parse_requirement_details(tokenizer)
+    tokenizer.expect("END", expected="end of dependency specifier")
+    return ParsedRequirement(name, extras, url, specifier, marker)
 
-def parse_named_requirement(requirement: str) -> Requirement:
+
+def _parse_requirement_details(
+    tokenizer: Tokenizer,
+) -> Tuple[str, str, Optional[MarkerList]]:
     """
-    named_requirement:
-        IDENTIFIER extras (URL_SPEC | specifier) (SEMICOLON marker_expr)? END
+    requirement_details = AT URL (WS SEMICOLON marker WS?)?
+                        | specifier WS? (SEMICOLON marker WS?)?
     """
-    tokens = Tokenizer(requirement)
-    tokens.expect("IDENTIFIER", error_message="Expression must begin with package name")
-    name = tokens.read("IDENTIFIER").text
-    extras = parse_extras(tokens)
+
     specifier = ""
     url = ""
-    if tokens.match("URL_SPEC"):
-        url = tokens.read().text[1:].strip()
-    elif not tokens.match("END"):
-        specifier = parse_specifier(tokens)
-    if tokens.try_read("SEMICOLON"):
-        marker = ""
-        while not tokens.match("END"):
-            # we don't validate markers here, it's done later as part of
-            # packaging/requirements.py
-            marker += tokens.read().text
+    marker = None
+
+    if tokenizer.check("AT"):
+        tokenizer.read()
+        tokenizer.consume("WS")
+        url_start = tokenizer.position
+        url = tokenizer.expect("URL", expected="URL after @").text
+        if tokenizer.check("END", peek=True):
+            return (url, specifier, marker)
+
+        tokenizer.expect("WS", expected="whitespace after URL")
```

Comment:
It seems that this resolves #529 (and therefore #592) too, could probably add closing keywords for them in the PR description :)

Reply:
Indeed!